### PR TITLE
[Docs] Add DR to cover default method implementation strategy

### LINKED ISSUE #1099: Document approach to default implementations
## What

Add a DR to capture and ratify our approach to default method implementations.

## Why

There are significant pros/cons to which ever route we take, we need to capture this to ensure we can inform future development.

## Acceptance Criteria

- Ratified DR merged that defines our strategy

### PR DESCRIPTION
**Pull request by foundrytom**
@elliotcmorris @feltech does this capture what we talked about? Whilst writing it a few revelations occurred - highlighted as questions in the intro.

Rendered version [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/1101/files?short_path=9af3a64#diff-9af3a643212c76abfa0cf12c698c4efc6e2757c13b5b59d2acbfce61e661131a) for 👁️ 

**Host authors, please share your thoughts! It affects you quite a lot!**

Closes #1099.

### CODE REVIEW

**Review by feltech — COMMENTED** [review]
Still pondering, but here's some initial small comments

**Review by LindaWHogg — COMMENTED** [review]
Here are some general thoughts.
- I don't think that an optional method not being implemented is a runtime error so it shouldn't be treated as such.
- If this is a common scenario, it would be nice to have introspection capabilities to help distinguish between no data and not implemented.
- `managementPolicy` seems to be a good candidate for querying whether an optional method is supported.
- It would be nice to have the same mechanism for read and write operations. This seems to rule out option 1.
- Having said that, and since this is not an error, maybe option 1 would be the cleanest?
- As above, as optional method not being implemented does not seem like an exception to me. Also, lots of `try / catch` clauses might be messy, which makes option 2 less attractive.

**Comment by ChrisBeckfordFoundry** [comment]
It sounds like option 1 only solves half the problem, since it doesn't work for write operations, and Option 3 only works for batch operations, I'm defaulting to Option 2. Raising a NotImplemented exception works as an all-purpose solution, and is what I'd expect from a function that's not been implemented (even if handling them with try{}catch{} can lead to code-bloat). 

Ideally, it would come with better introspection capabilities to protect from excessive exception handling and tweaking of the UX based on what's available (rather than a **try it and see** approach), but should still work as a stand-alone option too.

File: doc/decisions/DR022-Default-method-implementations.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+  API methods. It is more about classes of interaction. For example,
+  querying an entity trait set for read, informs what traits can be
+  resolved for that type of entity. Querying for write informs how the
+  manager wishes to handle publishing - affecting the use of `prefight`,
+  `resolve` and `register`.
+
+- `managementPolicy` traits could be added to the core API to cover
+  introspection of the support for additional/specific methods.
+
+- The use of `managementPolicy` before calling an API method is not
```
[the remark this answers is not in the document]
Comment (foundrytom) [reply doc/decisions/DR022-Default-method-implementations.md:63]: Its not (at least at present) something we tell hosts is mandatory before calling API methods. Eg. the simple resolver sample code doesn't bother with it.

File: doc/decisions/DR022-Default-method-implementations.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+```python
+def on_success(idx, result):
+    if result:
+        publish_to(result)
+    else:
+        ask_user_for_ref()
+
+def on_error(idx, error):
+    if error.code = kNotImplemented:
```
[the remark this answers is not in the document]
Comment (foundrytom) [reply doc/decisions/DR022-Default-method-implementations.md:236]: It was just for consistency with the other sketches, where the host doesn't mind if there is no default, but would want to report other errors to the user as something is wrong with the input.

File: doc/decisions/DR022-Default-method-implementations.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+#### Cons
+
+- Exceptions should be used for exceptional situations, and this one is
+  potentially very common.
+- Complicates host control flow and/or forces the use of `try`/`except`
+  at the call site where the exact reason for empty data is not critical
+  to control flows.
+
+### Option 3 - Error with a `kNotImplemented` BatchElementError
```
[the remark this answers is not in the document]
Comment (foundrytom) [reply doc/decisions/DR022-Default-method-implementations.md:114]: Yeah its definitely an option, this whole topic came up as with the current strategy (defaults) there is no way to tell the difference if they _do_ care, so it seems like any solution should make sure it doesn't just move that ambiguity to another place?

[the remark this answers is not in the document]
Comment (feltech) [reply doc/decisions/DR022-Default-method-implementations.md:114]: OK sounds reasonable

File: doc/decisions/DR022-Default-method-implementations.md [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+  at the call site, without an additional check of the capabilities
+  API. This may mask programming errors in cases where the host
+  intended to adapt their behaviour.
+- Unsuitable for write operations.
+
+### Option 2 - `NotImplemented` exception
+
+Default implementations throw a `NotImplemented` exception.
+
+#### Pros
```
[the remark this answers is not in the document]
Comment (foundrytom) [reply doc/decisions/DR022-Default-method-implementations.md:102]: Added [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/1101/commits/87b9ea372d27cf5389981ec6accbb5b6a5862b71)

**Comment by foundrytom** [comment]
A sketch for the proposed introspection method here: #1113

**Pull request merged**